### PR TITLE
add kustomize.sh wrapper

### PR DESCRIPTION
**Pull request by schweikert**
Add a wrapper for easier distribution of the kustomize binary.

### CODE REVIEW

**Comment by schweikert** [comment]
I signed the CLA

File: INSTALL.md [hunk 1]
```diff
@@ -31,3 +31,13 @@ To install from head with [Go] v1.10.1 or higher:
 ```
 go get github.com/kubernetes-sigs/kustomize
 ```
+
+## Installation using the kustomize.sh wrapper
+
+You can use the [kustomize.sh](scripts/kustomize.sh) script to
```
Comment (Liujingfang1) [inline INSTALL.md]: Can you add a command for using this script. Maybe similar to
`curl https://api.github.com/repos/kubernetes-sis/kustomize/scripts/kustomize.sh`

File: scripts/kustomize.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
+#!/bin/bash
```
Comment (Liujingfang1) [inline scripts/kustomize.sh]: Maybe change this file to `install_kustomize.sh`

**Comment by Liujingfang1** [comment]
A few questions about this script:
- Kustomize also has a windows version, can you also include that?
- This script hard coded version 1.0.6. Do you plan to add any mechanism to install any available version?

File: scripts/kustomize.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ARCH=""
+case "$UNAME" in 
+	"Darwin i386")
+		ARCH="darwin_amd64"
+		CHECKSUM_EXPECTED="$CHECKSUM_DARWIN"
+		sha256() {
+			shasum -a 256 $1 | awk '{print $1}'
+		}
+		;;
+	"Linux x86_64")
```
Comment (monopole) [inline scripts/kustomize.sh]: This is too specific.  My system returns "linux unknown".

File: INSTALL.md (hunk 1 above)
Comment (schweikert) [reply INSTALL.md]: Maybe there is a misunderstanding about the intention of this script: it is a wrapper script that is used in place of the kustomize binary, and not just an installer.

The idea is that you can commit this to your project git repo (say in myproject/tools/kustomize.sh), and then you can call it everywhere from there, for example also in ci/cd build pipelines.

Showing a curl command doesn't really help, because it's not something that you only execute once. You should just download the script, possibly modify it to your liking, and put it in your git repo.

File: scripts/kustomize.sh (hunk 2 above)
Comment (schweikert) [reply scripts/kustomize.sh]: As stated above: this is not a "installer". It is supposed to be called instead of the kustomize binary. It will cache the real binary and "exec" it.

**Comment by schweikert** [comment]
@Liujingfang1 Yes, I will also add support for Windows. Possibly using "git for windows", which provides a bash shell.

@monopole what does "uname -ps" say on your system?

File: scripts/kustomize.sh (hunk 3 above)
Comment (schweikert) [reply scripts/kustomize.sh]: Yes, I can reproduce this too. I will improve this part.

Comment (schweikert) [reply scripts/kustomize.sh]: It should be fixed now.

**Comment by schweikert** [comment]
Windows support now also added

**Comment by Liujingfang1** [comment]
Just released v1.0.7. Can you change the script to work with the new version?

**Comment by schweikert** [comment]
Cool, thanks for version 1.0.7. I have updated the script.

Maybe one day the script could be published as part of the release process, as an additional asset (so that you always have the right version there).

File: scripts/kustomize.sh [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		ARCH="linux_amd64"
+		CHECKSUM_EXPECTED="$CHECKSUM_LINUX"
+		sha256() {
+			sha256sum $1 | awk '{print $1}'
+		}
+		;;
+	MINGW*)
```
Comment (Liujingfang1) [inline scripts/kustomize.sh]: What is MINGW?

**Comment by monopole** [comment]
This script is intended to replace direct execution of a binary with a script that tries to download a hard-coded release of that binary to $HOME/.cache/kustomize, verifies the checksum, and runs it via process replacement.

The problem with bash installation scripts is that it's almost impossible to make everyone happy across operating systems and distros - the variety is a blessing/curse.  The audience for a tool can be expected to be very opinionated about how installation happens.

My objections to making this "official":

 - In the case of windows, it assumes bash, which is not yet a safe assumption
 - i (personally) don’t want to put this binary in $HOME/.cache
 - i (personally) don’t want CI/CD scripts to run a bash script doing magic instead of running a specific version of a specific binary.
 - the checksum thing isn’t a security win, since we always download from the repo directly.  I don't want to ever have to edit it, so i'll have to rely on someone else to do so, and then i have to download a new version of the script, which i'll just skip in favor of getting the binary directly.

How about adding operating system detection (as brief as possible) to the bash script snippet on INSTALL.md (which always pulls latest), and call it a day.    That snippet doesn’t do much other than download and rename, but it’s also simple enough to immediately read and copy/paste to a shell.

File: scripts/kustomize.sh (hunk 4 above)
Comment (schweikert) [reply scripts/kustomize.sh]: MINGW is a port of bash and other Unix/Linux tools to Windows. It's what "Git BASH" for Windows uses.

The script being a bash script and the fact that Windows doesn't ship with bash makes this for Windows a bit special. I have the feeling that it can still be useful however. Many of my colleagues use Git BASH.

**Comment by schweikert** [comment]
@monopole: I don't follow your argument about having checksums not being a security. Maybe this is because I imagine that the script would be used in a different way than you. What I imagined is this:

- Whoever decides to start using kustomize in a project, adds the wrapper to the project's git repository, where the yaml files reside, and puts instructions in the build pipeline to use the wrapper.
- Before committing the script to the repo, the person that adds or reviews the script can verify that the checksums are correct. This is done only once and you could say that the script is the trust anchor.
- The script is meant as an example and can be customized by the person adding the script. You can for example change the download URL so that an internal repository is used.
- Whenever the pipeline runs, you can be safe assuming that exactly the version that you defined is going to run.

Basically the script simplifies what you would otherwise need to do in a pipeline. It is much easier to just call the wrapper, instead of taking care of all these things yourself.

**Comment by schweikert** [comment]
About the fact that it is "almost impossible to make everyone happy across operating systems and distros":

The script only relies on "uname -ms", which seems to be quite stable across distros. I would say that the architecture-detection code is probably as portable as go itself. Did you find any linux host where it doesn't run?

**Comment by schweikert** [comment]
About the script being opinionated (regarding for example where to download the binary): yes, but I think that's ok because the script is simple and the idea is that it can be customized to your needs/liking.

**Comment by monopole** [comment]
IMO, the checksum related code isn't a security win because the script downloads the binary from the same place it got the checksum (so its only a consistency check, and years of downloading whole iso's i've never once been disappointed by tcp).  The lock-in to a specific version (and all its checksums on various OS's), plus the lack of bash on windows, the assumption of .cache, and the assumption that one would store this with one's YAML, etc. make this a very _specific_ workflow, out of many others, the most traditional being _download the binary you want to /usr/local/bin_.  Anyone able to adapt this script to their own needs would already have written something simpler on their own.

Your effort is appreciated, and suggest you publish the script in your own repo, and file an issue here referring to it.  If the issue gets, say, 3 or 4 upvotes, then we'll just take it.

Does that sound reasonable?

**Comment by schweikert** [comment]
The security win is that you can check, once, that the checksum is right on github (maybe even verify the signature, if the checksum file was signed...) and be certain that you always download exactly that version.

Anyway, it's fine by me if you don't want to include it. I'll just publish it somewhere else.
